internal/perms/grants.go

```diff
@@ -343,154 +393,219 @@ func Parse(scopeId, grantString string, opt ...Option) (Grant, error) {
-					}
-				case 1:
-					if !grant.hasActionOrSubaction(action.Create) &&
-						!grant.hasActionOrSubaction(action.List) {
-						return Grant{}, errors.NewDeprecated(errors.InvalidParameter, op, fmt.Sprintf("parsed grant string %q contains non-create or non-list action in a format that only allows these", grant.CanonicalString()))
+		}
+	}
+	// It's possible that there is no id in a grant. In that case we still need
+	// to validate it and build up the parsed Grant. We insert an empty ID value
+	// in this case; the code below will check that it's non-empty before
+	// running any ID-specific logic on it.
+	if len(grantIds) == 0 {
+		grantIds = []string{""}
+	}
+	for i, currId := range grantIds {
+		// Check for templated values ID, and substitute in with the authenticated
+		// values if so. If we are using a dummy user or account ID, store the
+		// original ID and return it at the end; this is usually the case when
+		// validating grant formats.
+		var origId string
+		if currId != "" {
+			if strings.HasPrefix(currId, "{{") {
+				id := strings.TrimSuffix(strings.TrimPrefix(currId, "{{"), "}}")
+				id = strings.TrimSpace(id)
+				switch id {
+				case "user.id", ".User.Id":
+					if opts.withUserId != "" {
+						grantIds[i] = strings.ToValidUTF8(opts.withUserId, string(unicode.ReplacementChar))
+					} else {
+						// Otherwise, substitute in a dummy value
+						origId = currId
+						grantIds[i] = "u_dummy"
 					}
-				case 2:
-					if !grant.hasActionOrSubaction(action.Create) || !grant.hasActionOrSubaction(action.List) {
-						return Grant{}, errors.NewDeprecated(errors.InvalidParameter, op, fmt.Sprintf("parsed grant string %q contains non-create or non-list action in a format that only allows these", grant.CanonicalString()))
+				case "account.id", ".Account.Id":
+					if opts.withAccountId != "" {
+						grantIds[i] = strings.ToValidUTF8(opts.withAccountId, string(unicode.ReplacementChar))
+					} else {
+						origId = currId
+						grantIds[i] = "acctoidc_dummy"
 					}
 				default:
-					return Grant{}, errors.NewDeprecated(errors.InvalidParameter, op, fmt.Sprintf("parsed grant string %q contains non-create or non-list action in a format that only allows these", grant.CanonicalString()))
+					fieldName := "ids"
+					if deprecatedId {
+						fieldName = "id"
+					}
+					return Grant{}, errors.NewDeprecated(errors.InvalidParameter, op, fmt.Sprintf("unknown template %q in grant %q value", currId, fieldName))
 				}
 			}
 		}
-		// This might be zero if output fields is populated
-		if len(grant.actions) > 0 {
-			// Create a dummy resource and pass it through Allowed and ensure that
-			// we get allowed.
-			acl := NewACL(grant)
-			r := Resource{
-				ScopeId: scopeId,
-				Id:      grant.id,
-				Type:    grant.typ,
+
+		// We don't need to do these twice as they don't depend on IDs; they
+		// also clear state such as actionsBeingParsed
```

Comment:
```suggestion
		// also clearly state such as actionsBeingParsed
```

Reply:
Nope, this was correct. They clear state in the grant struct :-)